tests/test_build.py

```diff
@@ -1,202 +1,129 @@
+
 
-path_tests = Path(__file__).parent.resolve()
-path_books = path_tests.joinpath("books")
-path_root = path_tests.parent
+def test_create(tmpdir, cli):
+    book = Path(tmpdir) / "new_book"
+    result = cli.invoke(commands.create, str(book))
+    assert result.exit_code == 0
+    assert book.joinpath("_config.yml").exists()
+    assert len(list(book.iterdir())) == 9
 
 
-def test_build_book(tmpdir):
+def test_build_from_template(tmpdir, cli):
     """Test building the book template and a few test configs."""
     # Create the book from the template
-    path = Path(tmpdir).joinpath("mybook").absolute()
-    run(f"jb create {path}".split())
-
-    # Ensure the book is created properly
-    assert path.joinpath("_config.yml").exists()
-
-    # Build the book
-    run(f"jb build {path}".split(), check=True)
-    path_html = path.joinpath("_build", "html")
-    assert path_html.joinpath("index.html").exists()
-    assert path_html.joinpath("intro.html").exists()
-
-    # Test custom config values
-    path_config = path_books.joinpath("config")
-    run(f"jb build {path_config}".split(), check=True)
-    html = path_config.joinpath("_build", "html", "index.html").read_text(
-        encoding="utf8"
-    )
+    book = Path(tmpdir) / "new_book"
+    _ = cli.invoke(commands.create, str(book))
+    build_result = cli.invoke(commands.build, str(book))
+    assert build_result.exit_code == 0
+    html = book.joinpath("_build", "html")
+    assert html.joinpath("index.html").exists()
+    assert html.joinpath("intro.html").exists()
+
+
+def test_custom_config(cli, build_resources):
+    books, _ = build_resources
+    config = books.joinpath("config")
+    result = cli.invoke(commands.build, str(config))
+    assert result.exit_code == 0
+    html = config.joinpath("_build", "html", "index.html").read_text()
     assert '<h1 class="site-logo" id="site-title">TEST PROJECT NAME</h1>' in html
     assert '<div class="sphinx-tabs docutils container">' in html
     assert '<link rel="stylesheet" type="text/css" href="_static/mycss.css" />' in html
     assert '<script src="_static/js/myjs.js"></script>' in html
 
 
-def test_toc_builds(tmpdir):
+@pytest.mark.parametrize("toc", ["_toc.yml", "_toc_startwithlist.yml"])
+def test_toc_builds(cli, build_resources, toc):
     """Test building the book template with several different TOC files."""
-    path_output = Path(tmpdir).joinpath("mybook").absolute()
-
-    # Regular TOC should work
-    p_toc = path_books.joinpath("toc")
-    path_toc = p_toc.joinpath("_toc.yml")
-    out = run(f"jb build {p_toc} --path-output {tmpdir} --toc {path_toc} -W".split())
-
-    # TOC with a single-item list should work
-    p_toc = path_books.joinpath("toc")
-    path_toc = p_toc.joinpath("_toc_startwithlist.yml")
-    out = run(f"jb build {p_toc} --path-output {tmpdir} --toc {path_toc} -W".split())
-
-    # TOC errors
-    p_toc = path_books.joinpath("toc")
-
+    books, tocs = build_resources
+    toc = str(tocs / toc)
```

Comment:
what is this dark magic?

Reply:
`build_resources` is the pytest fixture defined in conftest.py:

```python
@pytest.fixture()
def build_resources(tmpdir):
    src = Path(__file__).parent.resolve().joinpath("books").absolute()
    dst = tmpdir.join("books")
    shutil.copytree(src, dst)
    books = Path(dst)
    tocs = books / "toc"
    yield books, tocs
    shutil.rmtree(dst)
```

After unpacking the two `pathlib.Path` objects yielded by the fixture, we get `books` and `tocs`.  `tocs` replaces the formerly globally defined variable `p_toc`.   `toc = str(tocs / toc)` gives the full path to the toc being passed in through pytest parameterization, in this case `["_toc.yml", "_toc_startwithlist.yml"]`.  This is all happening in the tmpdir set up by the fixture for every test, so there's no chance of side effects in the local repo or across individual test runs.